src/drivers/imu/bmi088/BMI088_accel.cpp

```diff
@@ -0,0 +1,597 @@
+{
+    /* make another measurement */
+    measure();
+}
+
+void
+BMI088_accel::check_registers(void)
+{
+    uint8_t v;
+
+    if ((v = read_reg(_checked_registers[_checked_next])) !=
+            _checked_values[_checked_next]) {
+        _checked_bad[_checked_next] = v;
+
+        /*
+                  if we get the wrong value then we know the SPI bus
+                  or sensor is very sick. We set _register_wait to 20
+                  and wait until we have seen 20 good values in a row
+                  before we consider the sensor to be OK again.
+     */
+        perf_count(_bad_registers);
+
+        /*
+                  try to fix the bad register value. We only try to
+                  fix one per loop to prevent a bad sensor hogging the
+                  bus.
+     */
+        if (_register_wait == 0 || _checked_next == 0) {
+            // if the product_id is wrong then reset the
+            // sensor completely
+            write_reg(BMI088_ACC_SOFTRESET, BMI088_SOFT_RESET);
+            _reset_wait = hrt_absolute_time() + 10000;
+            _checked_next = 0;
+
+        } else {
+            write_reg(_checked_registers[_checked_next], _checked_values[_checked_next]);
+            // waiting 3ms between register writes seems
+            // to raise the chance of the sensor
+            // recovering considerably
+            _reset_wait = hrt_absolute_time() + 3000;
+        }
+
+        _register_wait = 20;
+    }
+
+    _checked_next = (_checked_next + 1) % BMI088_ACCEL_NUM_CHECKED_REGISTERS;
+}
+
+void
+BMI088_accel::measure()
+{
+    perf_count(_measure_interval);
+
+    if (hrt_absolute_time() < _reset_wait) {
+        // we're waiting for a reset to complete
+        return;
+    }
+
+    struct Report {
+        int16_t     accel_x;
+        int16_t     accel_y;
+        int16_t     accel_z;
+        int16_t     temp;
+    } report;
+
+    /* start measuring */
+    perf_begin(_sample_perf);
+
+    // Checking the status of new data
+    uint8_t status;
+    status = read_reg(BMI088_ACC_STATUS);
+    if (!(status & BMI088_ACC_STATUS_DRDY)) {
+        perf_end(_sample_perf);
+        perf_count(_duplicates);
+        _got_duplicate = true;
+        return;
+    }
+
+    _got_duplicate = false;
+
+    /*
+         * Fetch the full set of measurements from the BMI088 in one pass.
+         */
+    uint8_t index = 0;
+    uint8_t accel_data[8]; // Need an extra byte for the command, and an an extra dummy byte for the read (see section "SPI interface of accelerometer part" of the BMI088 datasheet)
+    accel_data[index] = BMI088_ACC_X_L | DIR_READ;
+
+    const hrt_abstime timestamp_sample = hrt_absolute_time();
+
+    if (OK != transfer(accel_data, accel_data, sizeof(accel_data))) {
+        return;
+    }
+
+    check_registers();
+
+    /* Extracting accel data from the read data */
+    index = 2; // Skip the dummy byte at index=1
+    uint16_t lsb, msb, msblsb;
+
+    lsb = (uint16_t)accel_data[index++];
```

Comment:
This readout and conversion of data can be prone to errors. Please could someone check this.